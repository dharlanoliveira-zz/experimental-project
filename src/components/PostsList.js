import React, {useEffect, useRef, useState} from "react";
import {useDispatch, useSelector} from 'react-redux';
import {postsActions} from "../actions/postsActions";
import Button from "@material-ui/core/Button";
import Grid from "@material-ui/core/Grid";
import TextField from "@material-ui/core/TextField";
import {ItemsList} from "./ItemsList"
import _ from "lodash"


function PostsList() {

    const [description,setDescription] = useState("")
    const dispatch = useDispatch()
    const posts = useSelector(state => {
        console.log("printing state")
        console.log(state)
        return state.posts.posts
    })

    useEffect(() => {
        dispatch(postsActions.getPosts())
    }, [])

    function renderNoPostsContent() {
        return (
            <>
                Don't exist any post
            </>
        )
    }

    function renderPostContent(post) {
        return (
            <>
                {post.description} - {post.language}
            </>
        )
    }

    function handleAddPost() {
        dispatch(postsActions.newPost(description))
        setDescription("")
    }

    function handleClearPost() {
        dispatch(postsActions.clearPosts())
        setDescription("")
    }

    return (
        <Grid container spacing={2}>
            <Grid item xs={12} md={12} lg={12}>
                    <ItemsList items={posts} emptyListRender={renderNoPostsContent}
                               listItemRender={renderPostContent}/>
            </Grid>

            <Grid item xs={12} md={12} lg={12}>
                <TextField type="text" name="description"
                           datatest-id="novoPost" aria-labelledby="novoPost"
                           label="Novo Post" value={description}
                           onChange={(event) => {
                                let { target: { value } } = event
                                setDescription(value)
                           }}/>
            </Grid>
            <Grid item xs={12} md={12} lg={12}>
                <Button id="adicionarPost" onClick={() => handleAddPost()} variant="contained" color="primary" aria-label="">
                    Add Post
                </Button>
                <Button id="adicionarPost"
                        onClick={() => handleClearPost()} variant="contained" color="primary"
                        disabled={_.isEmpty(posts)}>
                    Clear posts
                </Button>
            </Grid>
        </Grid>
    )
}

export {PostsList}