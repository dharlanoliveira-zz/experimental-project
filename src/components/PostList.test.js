import {PostsList} from "./PostsList";
import axios from 'axios';
import React from "react";
import {renderWithRedux, fireEvent, screen, getTextFieldByLabel, waitFor} from '../test/test-utils'

test('should add new posts list', async () => {
    let emptyPosts = {
        data: []
    };

    let onePost = {
        data: [{description: 'post1', language: 'pt'}]
    };

    let languageData = {
        data: {
            language: 'ptbr'
        }
    };

    axios.get
        .mockResolvedValueOnce(emptyPosts)
        .mockResolvedValueOnce(onePost)


    axios.post
        .mockResolvedValueOnce(languageData)

    const {debug} = renderWithRedux(<PostsList/>)
    debug()
    fireEvent.change(getTextFieldByLabel(screen, 'Novo Post'), {target: {value: '23'}})
    fireEvent.click(screen.getByText('Add Post'))
    await waitFor(() => expect(axios.post).toHaveBeenCalledTimes(1))
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(2))
    expect(screen.queryByText('post1 - pt')).toBeVisible()
});

test('should render posts list', async () => {
    let data = {
        data: [{description: 'post1', language: 'pt'}, {description: 'post2', language: 'en'}]
    };

    axios.get.mockResolvedValue(data)

    const {debug} = renderWithRedux(<PostsList/>)
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1))
    debug()
    await screen.findByText('post1 - pt')
    expect(screen.queryByText('post1 - pt')).toBeVisible()
    expect(screen.queryByText('post2 - en')).toBeVisible()
});


