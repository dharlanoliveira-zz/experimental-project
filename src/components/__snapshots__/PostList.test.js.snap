// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Items List Empty List 1`] = `
<div>
  <div
    className="MuiBox-root MuiBox-root-34"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      empty
    </p>
  </div>
</div>
`;

exports[`Storyshots Items List Twenty posts 1`] = `
<div>
  <div
    className="MuiBox-root MuiBox-root-183"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 0
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-214"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 1
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-215"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 2
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-216"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 3
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-217"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 4
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-218"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 5
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-219"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 6
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-220"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 7
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-221"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 8
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-222"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 9
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-223"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 10
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-224"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 11
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-225"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 12
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-226"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 13
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-227"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 14
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-228"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 15
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-229"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 16
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-230"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 17
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-231"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 18
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-232"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      description 19
    </p>
  </div>
</div>
`;

exports[`Storyshots Items List Two posts 1`] = `
<div>
  <div
    className="MuiBox-root MuiBox-root-99"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      content1
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-130"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      content2
    </p>
  </div>
</div>
`;

exports[`Storyshots Posts List Empty List 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12"
    >
      <div
        className="MuiBox-root MuiBox-root-561"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          Don't exist any post
        </p>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="standard-basic"
          id="standard-basic-label"
        >
          Novo Post
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="standard-basic"
            name="description"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        id="adicionarPost"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Add Post
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        disabled={true}
        id="adicionarPost"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Clear posts
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Storyshots Posts List Two posts 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12"
    >
      <div
        className="MuiBox-root MuiBox-root-1020"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          post1
           - 
          pt
        </p>
      </div>
      <div
        className="MuiBox-root MuiBox-root-1051"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          post2
           - 
          en
        </p>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="standard-basic"
          id="standard-basic-label"
        >
          Novo Post
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="standard-basic"
            name="description"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        id="adicionarPost"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Add Post
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        id="adicionarPost"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Clear posts
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Storyshots Posts list With some posts 1`] = `
<div
  style={
    Object {
      "width": "200px",
    }
  }
>
  <h3>
    Don't exist any post
  </h3>
  <input
    name="description"
    type="text"
  />
  <button
    id="adicionarPost"
    onClick={[Function]}
  >
    Add post
  </button>
</div>
`;

exports[`Storyshots Posts list default 1`] = `
<div
  style={
    Object {
      "width": "200px",
    }
  }
>
  <h3>
    Don't exist any post
  </h3>
  <input
    name="description"
    type="text"
  />
  <button
    id="adicionarPost"
    onClick={[Function]}
  >
    Add post
  </button>
</div>
`;
